src/typecheck/mod.rs

```diff
@@ -1077,9 +1085,41 @@ fn walk_with_annot<L: Linearizer>(
     }
 }
 
-/// Typecheck a term against a specific type.
+/// Check a term against a given type. Although this method mostly corresponds to checking mode in
+/// the classical bidirectional framework, it combines both checking and inference mode in
+/// practice, to avoid duplicating rules (that is, code) as detailed below.
+///
+/// # Literals
+///
+/// Checking a literal (a number, a string, a boolean, etc.) unifies the checked type with the
+/// corresponding primitive type (`Number`, `String`, `Bool`, etc.). If the checked type is a
+/// unification variable, `check` acts as an inference rule. If the type is concrete, unification
+/// enforces equality, and `check` acts as a checking rule.
+///
+/// # Introduction rules
 ///
-/// # Arguments
+/// Following Pfenning's recipe (see [Bidirectional Typing][bidirectional-typing]),
+/// introduction rules (e.g. typechecking a record) are checking. `check` follows the same logic
+/// here: it uses unification to "match" on the expected type (in the case of record, a record type
+/// or a dictionary type) and pushes typechecking down the record fields.
+///
+/// # Elimination rules
+///
+/// Elimination rules (such as function application or primitive operator application) are only in
+/// inference mode (still following Pfenning's recipe). `check` follows the inference mode here
+/// (typically on function application, where we first call to `infer` on the function part, and
+/// then check the argument).
+///
+/// Still, `check` is supposed to be implementing checking mode from the outside. We thus also
+/// apply the rule which switches from inference to checking mode. This is in particular where
+/// subtyping happens, for example, when there is subtyping. Currently, subtyping isn't supported
+/// yet in Nickel, so the corresponding switching rule doesn't do much. But the implementation of
+/// RFC004 will bring subtyping, and [`subsumption`] will be the place to apply it.
```

Comment:
This sentence confused me for a second. It looked like a leftover from editing "for example, when there is subtyping", but it makes standalone sense as well. My suggestion hopefully makes this clearer.
```suggestion
/// apply the rule which switches from inference to checking mode. This occurs for example where
/// subtyping happens, at least when there is subtyping. Currently, subtyping isn't supported
/// in Nickel, yet, so the corresponding switching rule doesn't do much. But the implementation of
/// RFC004 will bring subtyping, and [`subsumption`] will be the place to apply it.
```

Reply:
I agree the original wording is bad. However, my pet-peeve with this suggestion is that it kinda reverse the relation between the switch rule and subtyping: it makes it sound like subtyping (when there is) happens at some places, depending on the precise type system design, and those places are where we should also switch from inference mode to checking mode.

In fact, it is the converse: the switching rule is pre-existing and makes sense alone, and if you have subtyping and follows the bidirectional approach, the only place you should/can apply subytping is when switching.

I'll try to reword this part anyway.